influxdb_store.go

```diff
@@ -0,0 +1,331 @@
+}
+
+func (in *InfluxDBStore) Collect(id SpanID, anns ...Annotation) error {
+	// Current strategy is to remove existing span and save new one
+	// instead of updating the existing one.
+	// TODO: explore a more efficient alternative strategy.
+	if err := in.removeSpanIfExists(id); err != nil {
+		return err
+	}
+	// trace_id, span_id & parent_id are set as tags
+	// because InfluxDB tags are indexed & those values
+	// are uselater on queries.
+	tags := make(map[string]string, 3)
+	tags["trace_id"] = id.Trace.String()
+	tags["span_id"] = id.Span.String()
+	tags["parent_id"] = id.Parent.String()
+	// Saving annotations as InfluxDB measurement spans fields
+	// which are not indexed.
+	fields := make(map[string]interface{}, len(anns))
+	for _, ann := range anns {
+		fields[ann.Key] = string(ann.Value)
+	}
+	// InfluxDB point represents a single span.
+	pts := []influxDBClient.Point{
+		influxDBClient.Point{
+			Measurement: spanMeasurementName,
+			Tags:        tags,   // indexed metadata.
+			Fields:      fields, // non-indexed metadata.
+			Time:        time.Now(),
+			Precision:   "s",
+		},
+	}
+	bps := influxDBClient.BatchPoints{
+		Points:          pts,
+		Database:        dbName,
+		RetentionPolicy: "default",
+	}
+	_, err := in.con.Write(bps)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (in *InfluxDBStore) Trace(id ID) (*Trace, error) {
+	trace := &Trace{}
+	// GROUP BY * -> meaning group by all tags(trace_id, span_id & parent_id)
+	// grouping by all tags includes those and it's values on the query response.
+	q := fmt.Sprintf("SELECT * FROM spans WHERE trace_id='%s' GROUP BY *", id)
+	result, err := in.executeOneQuery(q)
+	if err != nil {
+		return nil, err
+	}
+	// result.Series -> A slice containing all the spans.
+	if len(result.Series) == 0 {
+		return nil, errors.New("trace not found")
+	}
+	var isRootSpan bool
+	// Iterate over series(spans) to create trace children's & set trace fields.
+	for _, s := range result.Series {
+		span, err := newSpanFromRow(&s)
+		if err != nil {
+			return nil, err
+		}
+		if span.ID.Parent == 0 && isRootSpan {
+			// Must be a single root span.
+			return nil, errors.New("unexpected multiple root spans")
+		}
+		if span.ID.Parent == 0 && !isRootSpan {
+			isRootSpan = true
+		}
+		annotations, err := annotationsFromRow(&s)
+		if err != nil {
+			return trace, nil
+		}
+		span.Annotations = *annotations
+		if isRootSpan { // root span.
+			trace.Span = *span
+		} else { // children span.
+			trace.Sub = append(trace.Sub, &Trace{Span: *span})
+		}
+	}
+	return trace, nil
+}
+
+func (in *InfluxDBStore) Traces() ([]*Trace, error) {
+	traces := make([]*Trace, 0)
+	// GROUP BY * -> meaning group by all tags(trace_id, span_id & parent_id)
+	// grouping by all tags includes those and it's values on the query response.
+	rootSpansQuery := fmt.Sprintf("SELECT * FROM spans WHERE parent_id='%s' GROUP BY * LIMIT %d", zeroID, in.tracesPerPage)
```

Comment:
yes, we can use  [offset](https://docs.influxdata.com/influxdb/v0.9/query_language/data_exploration/#paginate-query-returns-with-offset) to provide pagination.